---
title: "Orbicella Faveolata Reuptake"
author: "Maria Ingersoll, Jamie Poirer, Susritha Kopparapu"
date: "4/29/2021"
output:
  html_document: default
---


## Introduction

## Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### R Version
R version 4.0.3 was used for this analysis. 

#### Packages
The following packages were used to clean, analyze, and visualize data. Packages were installed with [Bioconductor](https://bioconductor.org/biocLite.R) version 3.12 or the [Cran Repository](http://cran.us.r-project.org).   
```{r, message=FALSE, warning=FALSE}
library("DESeq2")                   #Version 1.30.1
library("arrayQualityMetrics")      #Version 3.46.0
library("dplyr")                    #Version 1.0.5
library("RColorBrewer")             #Version 1.1.2
library("gplots")                   #Version 3.1.1
library("tidyverse")                #Version 1.3.0
library("pheatmap")                 #Version 1.0.12
library("vegan")                    #Version 2.5.7
library("ggplot2")                  #Version 3.3.3
library("ggrepel")                  #Version 0.9.1
```

```{r, eval=FALSE, echo=FALSE}
packageVersion("DESeq2")
packageVersion("arrayQualityMetrics")
packageVersion("dplyr")
packageVersion("RColorBrewer")
packageVersion("gplots")
packageVersion("tidyverse")
packageVersion("pheatmap")
packageVersion("vegan")
packageVersion("ggplot2")
packageVersion("ggrepel")


#packageVersion("affycoretools")
#packageVersion("genefilter")

```

## Raw Data Filtering

```{r, eval=FALSE}
setwd("/Users/mariaingersoll/Desktop/BU2020-2021/Ecol_and_Enviro_Genomics/BI586-git/Ofav_uptake_finalproj")
```

```{r, echo=FALSE}
setwd("~/BI 586/FinalProject")
```

Counts data was stored in a text file that was read in as a data frame to a variable called `countData`. Note that this data was pre-filtered for a base mean of 3. The column names are renamed based on the the strain type and replicate number for readability sake. The length of the data frame is reported and it represents the number of isoforms.
```{r}
countData <- read.table("orbfav_uptake_counts.txt")
newColNames = c("BR.A", "BR.B", "BR.C", "BR.D", "CR.A", "CR.B", "CR.C","DR.A","DR.B", "DR.C", "DR.D")
colnames(countData)=paste(newColNames)
length(countData[,1])
```

The total sum of isoform counts for each condition are displayed below in a barplot. The minimum and maximum are as reported below. 
```{r, echo=FALSE, fig.cap="**Figure 1.** All raw sequencing count values from each *Orbicella faveolata* sample demonstrating various degrees of library sizes. Green bars are *O. faveolata* hosting Symbiodiniaceae strain B01 (replicates: BR.A, BR.B, BR.C, BR.D). Red bars are *O. faveolata* hosting Symbiodiniaceae strain C (triplicates: CR.A,CR.B, CR.c). Blue bars are *O. faveolata* hosting Symbiodiniaceae strain D (replicates: DR.A, DR.B, DR.C, DR.D).", fig.width=10}
totalCounts=colSums(countData)
barplot(totalCounts, col=c("green",  "green", "green", "green", "red", "red", "red", "blue", "blue" , "blue" , "blue"), ylab="Raw Counts", xlab="Coral Hosting Symbiodiniaceae Strains")
```

```{r}
min(totalCounts)
max(totalCounts)
```

#### Remove small library size replicates

Due to the small library sizes, one replicate from the strain B, one replicate from the strain D conditions, and all replicates from O. faveolata hosting Symbiodiniaceae strain C were removed. `newCounts` is generated from the original `countData` but only selecting the samples that have larger library size and excluding the following ones as they are too small. 
```{r}
newCounts = countData %>% 
  dplyr::select(-BR.D, -CR.A, -CR.B, -CR.C, -DR.B)
```

The total sum of isoform counts for each condition are displayed below in a barplot. The minimum and maximum are as reported below. 
```{r, echo=FALSE, fig.cap="**Figure 2.** Selected raw sequencing count values from each *Orbicella faveolata* sample demonstrating various degrees of library sizes. Green bars are *O. faveolata* hosting Symbiodiniaceae strain B01 (triplicates: BR.A, BR.B, BR.C). Blue bars are *O. faveolata* hosting Symbiodiniaceae strain D (triplicates: DR.A, DR.C, DR.D). One replicate from the strain B condition, one replicate from the strain D condition, and all replicates from *O. faveolata* hosting Symbiodiniaceae strain C were removed prior to this analysis due to small library sizes."}
totalCounts=colSums(newCounts)
barplot(totalCounts, col=c("green", "green", "green", "blue" , "blue" , "blue"), ylab="Raw Counts", xlab="Coral Hosting Symbiodiniaceae Strains")
```

```{r}
min(totalCounts)
max(totalCounts)
```

*Note: In the above barplot, we see that despite the previous filtering, there are still conditions that have large library sizes (BR.C, DR.A, and DR.C) compared to other conditions that have very low library sizes (BR.A, BR.B, annd DR.B). This inconsistency in library size is the reason why the following rarefaction was performed.* 

#### Rarefy large sample size replicates

Because the difference between the minimum and maximum library size was so large, the data was rarefied and high counts were cut down to 100,000. To use the `rrarefy` function from the `vegan` package, the `newCounts` data frame variable had to be transposed with the `t` function. This transposed newCounts called `t_newCounts` data frame was supplied to `rrarefy` with a subsample size of 100,000. The new rarified counts are called `rare` and can be transposed again to provide the rarified counts used to generate the following bar plot. 

The rarefaction steps were only run once and the new counts were saved to a file. 
```{r, warning=FALSE, eval=FALSE}
t_newCounts = t(newCounts)
rare = rrarefy(t_newCounts, 100000)
unt_newCounts = t(rare)
write.table(unt_newCounts, file="rare_orbfav_uptake_counts.txt", quote=F, sep="\t")
```

The rarefaction counts are read in from the file. This is because every time `rrarfey` is run it randomly rarefies the replicates with counts greater than 100,000. This way the same counts are used for all further analysis. 
```{r}
unt_newCounts = read.table(file="rare_orbfav_uptake_counts.txt", sep="\t")
```

The total sum of isoform counts for each condition are displayed below in a barplot. The minimum and maximum are as reported below. 
```{r, echo=FALSE, fig.cap="**Figure 3.** *Orbicella faveolata* samples with raw sequencing count values greater than 100,000 were rarefied to a maximum of 100,000 total counts. Green bars are *O. faveolata* hosting Symbiodiniaceae strain B01 (triplicates: BR.A, BR.B, BR.C). Blue bars are *O. faveolata* hosting Symbiodiniaceae strain D (triplicates: DR.A, DR.C, DR.D). The samples that were rarefied were replicates BR.C, DR.A, and DR.D, while all other replicates remain as original."}
rare_totalCounts=colSums(unt_newCounts)
barplot(rare_totalCounts, col=c("green", "green", "green", "blue" , "blue" , "blue"), ylab="Raw Counts", xlab="Coral Hosting Symbiodiniaceae Strains")
```

```{r}
min(rare_totalCounts)
max(rare_totalCounts)
```

*NOTE: All following analysis will be performed with the rarefied data and it must be kept in mind for conclusions drawn about the differential expression of genes between coral hosting strain B and coral hosting strain D.*

## Outlier Analysis

For outlier analysis, we first create a character vector `treat` holding the names of the strains of Symbiodiniaceae that each coral sample is hosting. The vector is converted to a dataframe and is renamed `colData` for further use. 
```{r}
treat=c( "B", "B", "B", "D", "D", "D")
g=data.frame(treat)
colData<- g
```

`DESeqDataSetFromMatrix` was run on the `unt_newCounts` which represents the rarefied counts of the selected samples. The column data is set to the strain type and the design to the treatment vector so that it may dictate how the counts for each gene depend on the variables in colData. The `DESeq` step was run on the results of the prior step.
```{r results="hide", message=FALSE, warning=FALSE, eval=TRUE}
dds<-DESeqDataSetFromMatrix(countData=unt_newCounts, colData=colData, design=~treat)
dds<-DESeq(dds)
```

The `vst` function is the variance-stabilizing transformation which models the mean-variance relationship of replicates. The assay of `dds` generates a table with the mean-variance of each of the isoform treatments. The `as.data.frame(colData(rl)))` generates a table with the columns and a sizeFactor value that is converted into the formal `AnnotatedDataFrame` class. Both values are passed are used to create an `ExpressionSet` object which is used to generate the quality metrics with `arrayQualityMetrics` using the treat variable as the intgroup.

```{r, eval=FALSE}
vsd.ge=assay(vst(dds))
rl=vst(dds)
e=ExpressionSet(assay(rl), AnnotatedDataFrame(as.data.frame(colData(rl))))
v = setwd("/Users/mariaingersoll/Desktop/BU2020-2021/Ecol_and_Enviro_Genomics/BI586-git/Ofav_uptake_finalproj")
arrayQualityMetrics(e,outdir=v,intgroup=c("treat"),force=T)
```
*NOTE: The array quality metrics showed that there were no outliers by the three detection methods: by distances between arrays, by boxplots, and by MA plots. Below is the heatmap of sample distance between arrays.*


`results` extracts the results table from the DESeq analysis and the `DESeqDataSet` is used to generate the dispersion plot.
```{r, warning=FALSE, message=FALSE, eval=TRUE}
head(dds)
res<- results(dds)
```

```{r, echo=FALSE, fig.cap="**Figure 4.** Scatter plot of dispersion estimates per gene (on the y-axis) versus the mean of normalized counts (on the x-axis) of isoforms identified in the treatment samples of *Orbicella faveolata*. Gene-wise dispersion estimates are in black, fitted estimates are in red, and final estimates are in blue as indicated by the legend.", eval=TRUE}
plotDispEsts(dds, main="Dispersion plot Uptake")
```

R log transformation is used on the data `dds` as a data normalization method. The rlog data provides a better transformation when size factors vary across samples by transforming to a log2 scale to minimize difference between samples for rows with small counts. This r log tranformation will be important for making heatmaps.
```{r, eval=TRUE}
rld <- rlogTransformation(dds, blind=TRUE)
```

The rlog data `rld` can be converted to a matrix like object with `assay` that is then given to the `hist` function to generate a histogram of the r-log transformations. Since the rlog values can be positive or negative, the histogram shows the spread of those positive and negative values for all the samples. 
```{r, echo=FALSE, fig.cap="**Figure 5.** A histogram of the binned r log transformations frequency of *Orbicella faveolata* data where the regularized log (r log) is the original count data on a log2 scale.", eval=TRUE}
hist(assay(rld))
```


The matrix like object of the r log data `rld` is transposed and the `dist` function is performed on it to calculate the distance matrix using the specified distance measure to compute the distance between the rows. This is then stored in the `sampleDists` variable. 
```{r results="hide", message=FALSE, eval=TRUE}
sampleDists <- as.matrix(dist(t(assay(rld))))
```

The `sampleDists` matrix is used to generate a sample distance heatmap of relatedness based on the count distance or difference between samples as seen below.
```{r, echo=FALSE, fig.cap="**Figure 6.** A sample distance heatmap created by calculating the distances between samples based on overall differences in the logfoldchange of all genes. Note that the *Orbicella faveolata* hosting Symbiodiniaceae strain B01 sample BR.C is clustering more with *O.faveolata* hosting Symbiodiniaceae strain D.", eval=TRUE}
heatmap.2(as.matrix(sampleDists), key=F, trace="none",
          col=colorpanel(100, "black", "white"),
          margin=c(10, 10))
```

The `results` function is applied to the `dds` data using a contrast by treat where B serves as the control since it is listed as the second term. The results are stored in a variable called `resBD`. We isolate the column of the data set for the p-adjusted value, and below is a report of how many of the isoforms have a p-adjusted value less than 0.1. Additionally, a summary of the `resBD` is generated as well to analyze the count differences between B and D. 
```{r, eval=TRUE}
resBD <- results(dds, contrast=c("treat","D","B"))
table(resBD$padj<0.1)
summary(resBD)
```

In the rarefied data, with an adjusted p-value of <0.1, 4 genes were significantly upregulated, and 26 were significantly downregulated in coral hosting strain D compared to coral hosting strain B. 

```{r, echo=FALSE, fig.cap="**Figure 7.** *Orbicella faveolata* hosting Symbiodiniaceae strain B01 vs *O.faveolata* hosting Symbiodiniaceae strain D scatter plot of log fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis) of the isoforms identified in the samples. The non significant data points are in gray, and significant data points are represented in blue. A positive log fold change indicates that D is upregulated compared to B01, and a negative log fold change indicates that D is downregulared compared to the baseline of B01.", eval=TRUE}
plotMA(resBD, main="O.faveolata hosting strain B01 vs O.faveolata hosting strain D")
```

A table of the log fold change values between coral hosting strain B and coral hosting strain D is generated. The file contains the log fold change, the p value, and the adjusted p value. 
```{r, eval=FALSE}
write.table(resBD, file="rare_BD_DE.txt", quote=F, sep="\t")
```

## P-values and R-log data

A new table `valBD` is generated from the p-value and adjusted p-value from the original `resBD` table. The column names are appropriately set and the number of isoforms are confirmed to be the same as the original dataset, no data was lost. 
```{r, eval=TRUE}
valBD=cbind(resBD$pvalue, resBD$padj)
colnames(valBD)=c("pval.BD", "padj.BD")
length(valBD[,1])
```


A table of the number of complete cases where both the p-value and the adjusted p-value is present and not missing are reported. 
```{r, eval=TRUE}
table(complete.cases(valBD))
```

The `rld` is the variable that holds the results of the `rlogTransformation` on the `dds` data where blind=True. The r log data is converted to a matrix like object with `assay` and the column names which were originally the names of each replicate are now converted to just the name of the strain that each coral sample is hosting. The number of isoforms are confirmed to be the same as the original dataset, and no data was lost. 
```{r, eval=TRUE}
rld=assay(rld)
colnames(rld)=paste(colData$treat)
length(rld[,1])
```

The r log transformation data is combined with the p-value and the adjusted p-value that to generate the new table `rldpvals` with the dimensions reported below. A table of the number of complete cases where all values for each are isoform is present and not missing are reported. 
```{r, eval=TRUE}
rldpvals=cbind(rld,valBD)
dim(rldpvals)
table(complete.cases(rldpvals))
```

The `rldpvals` are stored in a csv file for future use. 
```{r, eval=FALSE}
write.csv(rldpvals, "rare_BI586_uptake_RLDandPVALS.csv", quote=F)
rldpvals <- read.csv(file="rare_BI586_uptake_RLDandPVALS.csv", row.names=1)
```

## PCA Analysis 

To perform the Principal Component Analysis (PCA) on the data will allow us to visualize overall effect of experimental covariates and batch effects. We first must transpose the data table hols the r log transformation data assay. 
```{r,eval=TRUE}
rld_t=t(rld)
```

The `prcomp` function performs a principal component analysis on the transformed r log data matrix where the variables should be shifted to be zero centered and that any data that contains NA should be omitted from the analysis. The an instance of the `prcomp` class is returned and stored in a variable called `pca`.
```{r warning=FALSE, eval=TRUE}
pca <- prcomp(rld_t,center = TRUE, na.action=na.omit)
```

`sdev` holds the standard deviations of the principal components and it can be used to calculate the proportion that each principal component corresponds to the variance. We can then round the PC1 and PC2 (which are the first and second items of the list) to one significant figure. 
```{r,eval=TRUE}
li <- pca$sdev^2 / sum(pca$sdev^2)
pc1v <- round(li[1] * 100, 1)
pc2v <- round(li[2] * 100, 1)
```


The x column of the `prcomp` object holds the value of the rotated data, and this is all stored in the variable `pca_s`.
```{r,eval=TRUE}
pca_s <- as.data.frame(pca$x)
```

Since we only care about the first two dimensions, of PC1 and PC2, only the first two rotated data values from `pca_s` are needed. New columns for samples and treatment names are added to the `pca_s` table. 
```{r,eval=TRUE}
pca_s <- pca_s[,c(1,2)]
head(pca_s)
pca_s$Samples = row.names(pca_s)
pca_s$treat=colData$treat
```

Then the PCA plot can be generated from the PC1 and PC2 levels. 
```{r, echo=FALSE, fig.cap="**Figure 8.** Principal component analysis of all samples in the subset of data where the PC1 along the x-axis has a 35.3% variance and the PC2 along the y-axis has a 31% variance. The legend shows the difference between *Orbicella faveolata* hosting Symbiodiniaceae strain B01 vs *O.faveolata* hosting Symbiodiniaceae strain D  samples. Adonis below shows that PC1 is significant with an alpha of 0.1, and this can be considered trending towards significance.", eval=TRUE}
cbPalette <- c("darkorchid4","firebrick4")
ggplot(pca_s, aes(PC1, PC2, color = treat, pch = treat)) +
  geom_point(size=3) +
  #  geom_text_repel(aes(label=Samples)) +
  scale_colour_manual(values=cbPalette)+
  theme_bw() +
  # geom_density2d(alpha=.5)+
  geom_polygon(alpha=.2)+
  xlab(paste0("PC1: ",pc1v,"% variance")) +
  ylab(paste0("PC2: ",pc2v,"% variance")) 
```

The PCA plot shows that samples of coral hosting strain B have a very high variance, however there is no overlap between the samples by treatment. 

We then perform the `adonis` function on the rotated pca data with the treatment as the formula and the `prcomp` matrix as the data. The method for this adonis is euclidean and we select to remove any data with NA. The adonis will show whether the conditions are significantly different. 
```{r, warnings=FALSE, message=FALSE, echo=FALSE, eval=TRUE}
adonis(pca$x ~ treat, data = pca_s, method='eu', na.rm = TRUE)
```

The `adonis` results show that PC1 is significant with an alpha of 0.1, and this can be considered trending towards significance where PC1 is explaining 35% of the variance between the two samples. 

## Pairwise Comparison of Coral Hosting Strain B vs Strain D
The treat column is represented as a factor with two levels being coral hosting strain D or strain B, where coral hosting strain B serve as the control. These are stored in a new column called D Then the `results` function extracts the results table of `DESeqDataSet` with the contrast set as treat, between strain D and strain B
```{r}
colData$D<-factor(colData$treat, levels=c("D","B"))
resD <- results(dds, contrast=c("treat","D","B"))
```

The results table reports the breakdown of how many genes had an adjusted p-value below 0.01, to be considered significant. The summary of the results with a contrast by treat (strain) is also displayed below. Notice that while only 4 of genes were upregulated, 26 were downregulated.

```{r}
table(resD$padj<0.01)
summary(resD)
```

The number of genes that have an adjusted p-value less than 0.05 and that do not have NA values for the adjusted p-values, are counted and reported below.

```{r}
nrow(resD[resD$padj<0.05 & !is.na(resD$padj),]) 
```

```{r, echo=FALSE, fig.cap="**Figure 9.** *Orbicella Faveolata* hosting Symbiodiniaceae strain B01 vs. *O. faveolata* hosting Symbiodiniaceae strain D  scatter plot of log fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis) of the isogroups identified in the treatments samples. The non signicant data points are in gray, and significant data points are represented in blue. "}
plotMA(resD, main="B vs D", ylim=c(-2,2))
```

```{r include=FALSE}
results <- as.data.frame(resD)
```

The number of upregulated and downregulated genes (respectively) that have an adjusted p-value less than 0.1 and that do not have NA values for the adjusted p-value, are counted and reported below.
```{r}
nrow(resD[resD$padj<0.1 & resD$log2FoldChange > 0 & !is.na(resD$padj),])
nrow(resD[resD$padj<0.1 & resD$log2FoldChange < 0 & !is.na(resD$padj),])
```


A table of the results with a contrast by treat between coral hosting strain B and coral hosting strain D is generated. The file is read into a variable called `cd` that will be used to genereate the GO table later on. 
```{r,eval=TRUE}
write.table(resD, file="rare_D_2021.txt", quote=F, sep="\t")
cd <- read.table("rare_D_2021.txt")
```

## GO table for MWU

Use the results from the DESeq analysis with the contrast as the temperature to generate a term called `go_input_D` that has the log p-value as a negative number if the log fold change is less than 0 and a positive number if the log fold change is greater than 0. The generated GO table for the specific isogroups in the samples will be used for the GO enrichment analysis later on.
```{r,eval=TRUE}
go_input_D = cd %>%
  tibble::rownames_to_column(var = "iso") %>%
  mutate(gene = gsub("ofav*", "isogroup", iso)) %>%
  mutate(mutated_p = -log(pvalue)) %>%
  mutate(mutated_p_updown = ifelse(log2FoldChange < 0, mutated_p*-1, mutated_p*1)) %>%
  na.omit() %>%
  select(gene, mutated_p_updown)

colnames(go_input_D) <- c("gene", "pval")

write.csv(go_input_D, file="rare_D_GO.csv", quote=F, row.names=FALSE)
```

## GO analysis 
Here we will do a preliminary GO enrichment analysis of all the genes. A GOMWU performs rankings and gives the enrichment of GO terms in the ranked list. The GO terms that show up may not actually contain only genes that are significantly differentially expressed and it is just the gene set that is just enriched. 

There are 3 sub-ontologies to plot dendograms of genes between coral hosting strain B and coral hosting strain D. 

- MF: Molecular function
- BP: Biological process
- CC: Cellular component

We read in the mutated log fold change p-values from the csv file and store it in the variable `BD_DE_GO`. Just a reminder that positive log fold changes are when genes are upregulated in strain D compared to strain B and a negative log fold change are when genes are downregulated in strain D compared to strain B. 
```{r,eval=TRUE}
BD_DE_GO = read.csv("rare_D_GO.csv")
```

Some notes before the dendograms are generated:

- The BD_DE_GO.csv file has two columns of gene id and teh continuous measure of significance. To perform the standard GO enrichment analysis based on Fisher's exact test, we use binary measure (0 or 1, i.e., either significant or not).
- The orb_fav_iso2go.txt file is tab-delimited and has one line per gene. Multiple GO terms are listed and separated by semicolon. 
- The *Orbicella faveolata* isogroup GO annotations are from Rachel Wright. 
- The GO database was downloaded from [Gene Ontology](http://www.geneontology.org/GO.downloads.ontology.shtml) and here the go_basic.obo file is being used, but was renamed go.obo. 

### MF (Molecular Function)
gomwuStats settings were specified with the largest so a GO category will not be considered if it contains more than 0.1 of the total number of genes, the smallest is specified so that a GO category should contain a minimum of 5 genes to be considered, and the clusterCutHeight is the threshold for merging gene-sharing terms.
```{r,eval=TRUE}
#head(read.csv("rare_D_GO.csv"))
input="rare_D_GO.csv"
goAnnotations="orb_fav_iso2go_new.tab"
goDatabase="go.obo"
goDivision_1="MF"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision_1,
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25)
```


As we can see, this new rarefied data against this transcriptome results in 0 GO terms at 10% FDR for the Molecular Function sub-ontology. Due to this, there is no dendogram to be generated. 

### BP (Biological Process)
gomwuStats settings were specified with the largest so a GO category will not be considered if it contains more than 0.1 of the total number of genes, the smallest is specified so that a GO category should contain a minimum of 5 genes to be considered, and the clusterCutHeight is the threshold for merging gene-sharing terms. 
```{r,eval=TRUE}
input="rare_D_GO.csv"
goAnnotations="orb_fav_iso2go_new.tab"
goDatabase="go.obo"
goDivision_2="BP"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision_2,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25)
```

As we can see, this new rarefied data against this transcriptome results in 2 GO terms at 10% FDR for the Molecular Function sub-ontology. Since there are some GO terms, we can now create a dendogram. 

To plot the dendrogram the settings were chosen as follows. The level1=0.1 is the FDR threshold for plotting where all GO categories are plotted containing genes exceeding the absValue.Level2=0.05 is the FDR cutoff to print in regular (not italic) font. Level3=0.01 is the FDR cutoff to print in large bold font. The text size and treeHeight are for the spacing and look of the actual dendrogram.
```{r, echo = FALSE, message=FALSE, warning=FALSE, eval=TRUE, fig.height=2, fig.cap="**Figure 10.**GO MWU plot with level 1 at 0.1, level 2 at 0.05, and level 3 at 0.01. The biological processes (BP) sub-ontology was used for this analysis. Blue represents GO terms that were downregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata*hosting Symbiodiniaceae strain B01."}
resultsBP=gomwuPlot(input,goAnnotations,goDivision_2,
                  absValue=1,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.5,
                  treeHeight=0.5,
                  colors=c("dodgerblue2","firebrick1","skyblue","lightcoral")
)
```

In general, red is upregulated in coral hosting strain D and blue is downregulated in coral hosting strain D (or upregulated in coral hosting strain B). However, both GO terms here were downregulated in coral hosting strain D compared to the baseline of coral hosting strain B. 

### CC (Cellular Components)
gomwuStats settings were specified with the largest so a GO category will not be considered if it contains more than 0.1 of the total number of genes, the smallest is specified so that a GO category should contain a minimum of 5 genes to be considered, and the clusterCutHeight is the threshold for merging gene-sharing terms.
```{r,eval=TRUE}
input="rare_D_GO.csv"
goAnnotations="orb_fav_iso2go_new.tab"
goDatabase="go.obo"
goDivision_3="CC"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision_3,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25)
```

As we can see, this new rarefied data against this transcriptome results in 11 GO terms at 10% FDR for the Molecular Function sub-ontology. Since there are some GO terms, we can now create a dendogram. 

To plot the dendogram the settings were chosen as follows. The level1=0.1 is the FDR threshold for plotting where all GO categories are plotted containing genes exceeding the absValue.Level2=0.05 is the FDR cutoff to print in regular (not italic) font. Level3=0.01 is the FDR cutoff to print in large bold font. The text size and treeHeight are for the spacing and look of the actual dendogram.

```{r, echo=FALSE, warning=FALSE, eval=TRUE, fig.height=4, fig.cap="**Figure 11.** GO MWU plot with level 1 at 0.1, level 2 at 0.05, and level 3 at 0.01. The cellular components (CC) sub-ontology was used for this analysis. Red represents GO terms that were upregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata* hosting Symbiodiniaceae strain B01 and blue represents GO terms that were downregulated in *O.faveolata* hosting strain D compared to *O.faveolata* hosting strain B01. "}
resultsCC=gomwuPlot(input,goAnnotations,goDivision_3,
                  absValue=1,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.5,
                  treeHeight=0.5,
)
```

Red is upregulated in coral hosting strain D and blue is downregulated in coral hosting strain D (or upregulated in coral hosting strain B). Here we see some terms can be considered cellular structure elements that were downregulated in coral hosting strain D where the baseline is coral hosting strain B. However, in this case there are some mitochondrial terms that were upregulated in coral hosting strain D compared to those hosting strain B. 

## Investigating DEGs in specific GO categories
Code in this section was adapted from [Dan Wuitchik]( https://github.com/wuitchik/Divergent-thermal-challenges-elicit-convergent-stress-signatures-in-aposymbiotic-Astrangia-poculata/blob/master/Astrangia_HotCold.pdf)

Based on the original GO analysis in the CC sub ontology, there was evidence that mitochondrial related elements were upregulated in coral hosting strain D compared to coral hosting strain B. To further analyze this data, the "mitochondrial membrane part" [CC GO term: GO:0044455], and "mitochondrial part" [CC GO term: GO:0044429] were focused on. 

`iso2go` was set up with the ofav gene id and all of its associated GO terms after being read in. 
```{r, eval=TRUE}
iso2go = read.table("orb_fav_iso2go_new.tab", sep="\t")
colnames(iso2go) <- c("gene_ID", "GO.terms")
```


A character vector of gene ID and description is made and the `iso2gene` is read in. The character vector is assigned to be the column names of the iso2gene data frame. Then the gene symbol is obtained from the descriptions that are in the file and the variable `gene` now contains the gene ID, full description and the gene symbol 
```{r,eval = TRUE}
newnames = c("gene_ID", "description" )
iso2gene = read.table("orb_fav_iso2gene_new.tab", sep="\t")
colnames(iso2gene) <- newnames

gene = iso2gene %>%
       mutate(gene_symbol = gsub(".* GN=", "", description)) %>%
       mutate(gene_symbol = gsub(" .*", "", gene_symbol))
```


### All DEG - Mitochondria
The r-log fold change values and p-values from prior are read in and stored in a variable called `BD_DE`. It is important to note that in this file the gene IDs are listed with ofav instead of isogroup. It is important to change this so that they can be linked in future steps when creating the heatmaps. The `rlog_BD` now contains the isoform names that are in the gene_ID column as well as the log fold change for each of the samples. By selecting for only the first 7 columns, the adjusted p-values and p-values are dropped and only the r log data remains in the variable. 
```{r,eval=TRUE}
BD_DE = read.csv("rare_BI586_uptake_RLDandPVALS.csv")
BD_DE = BD_DE %>%
  mutate(gene_ID = gsub("ofav*", "isogroup", X)) %>%
  select(-X) %>%
  relocate(gene_ID, .before=B) 
rlog_BD = BD_DE[,1:7]
```


The GO csv is read in and the column names are renamed as needed to confirm that they match the case of the column names in the other files. Then a variable containing all the genes that correspond to the GO terms of interest is created. The genes that are selected for and correspond to the parent GO terms are as follows: Mitochondrial Membrane Part (GO:0044455) and Mitochondrial Part (GO:0044429) in CC sub-ontology. Any genes with NA are dropped from the selection and data is stored in the `mitochondria` variable. 
```{r, warning=FALSE, message=FALSE, eval=TRUE}
cc = read.csv("CC_rare_D_GO.csv", sep="\t")
colnames(cc)[4]="gene_ID"
mitochondria = cc %>%
  filter(str_detect(term, "GO:0044455|GO:0044429")) %>%
  left_join(rlog_BD) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-name, -term, -lev, -value, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
```


The number of genes that correspond to the parent GO terms is counted and reported below. 
```{r, eval=TRUE}
nrow(mitochondria)
```

A total of 55 terms are reported to correspond with the parent GO terms for mitochondrial function and a heatmap is plotted. 
```{r, echo=FALSE, warning=FALSE, message=FALSE, eval=TRUE, fig.height=10, fig.cap="**Figure 12.** A heatmap of the differential gene expression of all genes that correspond to mitochondrial membrane function. Yellow/white represents genes that were upregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata* hosting Symbiodiniaceae strain B01 and red represents genes that were downregulated in *O.faveolata* hosting strain D compared to *O.faveolata* hosting strain B01."}
heatmap_mit = heatmap.2(as.matrix(mitochondria), Rowv = TRUE, Colv = FALSE, scale = "row",
          dendrogram = "both",
          trace = "none",
          main = "GO Mitochondria",
          margin = c(5,15))
```


### Significant (p-value < 0.1) DEG - Mitochondria
The uptake data is filtered for only those genes that were significantly differentially expressed (alpha=0.1). The `rlog_BD_10` now contains the isoform names that are in the gene_ID column as well as the log fold change for each of the samples. By selecting for only the first 7 columns, the adjusted p-values and p-values are dropped and only the r log data remains in the variable. The number of significantly expressed genes are counted and reported below. 
```{r,eval=TRUE}
BD_DE_10 = filter(BD_DE, pval.BD <= 0.1, preserve = TRUE)
rlog_BD_10 = BD_DE_10[,1:7]
nrow(rlog_BD_10)
```

We see that this is a total of 562 genes in this rarefied data for significantly differentially expressed genes where the alpha = 0.1. 


Now that the list has been filtered to only significantly differentially expressed genes, all genes that correspond to the GO terms of mitochondrial membrane function can be selected for and stored in `GO_mit_sig`. The number of significantly differentially expressed genes that are associated with the mitochondrial membrane function terms are counted and reported below. 
```{r,eval=TRUE, message=FALSE, warning=FALSE}
GO_mit_sig = cc %>%
  filter(str_detect(term, "GO:0044455|GO:0044429")) %>%
  left_join(rlog_BD_10) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-name, -term, -lev, -value, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
nrow(GO_mit_sig)
```


There are 5 such genes and the calculated differences from the means are plotted in a heatmap. 
```{r,eval=TRUE}
GO_mit_means_sig = apply(GO_mit_sig, 1, mean)
explc_mit = GO_mit_sig-GO_mit_means_sig
```


```{r, echo=FALSE, fig.height=5, fig.cap="**Figure 13.** A heatmap of the significantly differntially expressed genes (alpha = 0.1) that correspond to mitochondrial membrane GO terms generated from the calculated differences from the means. Orange represents genes that were upregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata* hosting Symbiodiniaceae strain B01 and blue represents genes that were downregulated in *O.faveolata* hosting strain D compared to *O.faveolata* hosting strain B01."}
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
heatmap_GO_mit_sig = heatmap.2(as.matrix(explc_mit), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                               dendrogram = "both",
                               trace = "none",
                               main = "GO Mitochondrial Membrane",
                               margin = c(5,15))
```


### Filter data for different alpha levels
The uptake data is filtered for only those genes that were significantly differentially expressed with stricter bounds for p-values (alpha=0.05 and alpha=0.01) as well as adjusted p-values (alpha = 0.1) to account for the count differences. The `rlog_BD_05`, `rlog_BD_01`, `rlog_BD_adj` now contain the isoform names that are in the gene_ID column as well as the log fold change for each of the samples. By selecting for only the first 7 columns, the adjusted p-values and p-values are dropped and only the r log data remains in the variable. The number of significantly expressed genes are counted and reported below. 
```{r,eval=TRUE}
BD_DE_05 = filter(BD_DE, pval.BD <= 0.05, preserve = TRUE)
rlog_BD_05 = BD_DE_05[,1:7]
nrow(rlog_BD_05)

BD_DE_01 = filter(BD_DE, pval.BD <= 0.01, preserve = TRUE)
rlog_BD_01 = BD_DE_01[,1:7]
nrow(rlog_BD_01)

BD_DE_adj = filter(BD_DE, padj.BD <= 0.1, preserve = TRUE)
rlog_BD_adj = BD_DE_adj[,1:7]
nrow(rlog_BD_adj)
```

- We see that this is a total of 279 genes in this rarefied data for significantly differentially expressed genes where the alpha = 0.05 for the p-value. 

- We see that this is a total of 68 genes in this rarefied data for significantly differentially expressed genes where the alpha = 0.01 for the p-value. 

- We see that this is a total of 30 genes in this rarefied data for significantly differentially expressed genes where the alpha = 0.1 for the *adjusted* p-value. 

### Significant (adjusted p-value < 0.1) DEG - Mitochondria
Now that the list has been filtered to only significantly differentially expressed genes with an adjusted p-value at alpha = 0.1, all genes that correspond to the GO terms of mitochondrial membrane function can be selected for and stored in `sigDEG_BD_adj`. The number of significantly differentially expressed genes that are associated with the mitochondrial membrane function terms are counted and reported below. 
```{r,warning=FALSE, message=FALSE, eval=TRUE}
sigDEG_BD_adj = iso2go %>%
  left_join(rlog_BD_adj) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
nrow(sigDEG_BD_adj)
```

Here there are only 18 of the 30 DEGs that had annotated GO terms. So a heatmap with these raw values can be generated. 

```{r,eval=TRUE, echo=FALSE,fig.width=9, fig.cap="**Figure 14.** A heatmap of the significantly differntially expressed genes (adjusted p-value where alpha = 0.1) that correspond to mitochondrial membrane GO terms. Orange represents genes that were upregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata* hosting Symbiodiniaceae strain B01 and blue represents genes that were downregulated in *O.faveolata* hosting strain D compared to *O.faveolata* hosting strain B01 "}
heatmap_sigDEG_adj = heatmap.2(as.matrix(sigDEG_BD_adj), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG, padj < 0.1",
                           margin = c(5,15))
```


Noting that many of the terms are un-annotated, a heatmap can be generated by calculating the difference from the means of the adjusted p-values of the differentially expressed genes. 
```{r,eval=TRUE}
sigDEG_BD_means_adj = apply(sigDEG_BD_adj, 1, mean)
explc_sig_adj = sigDEG_BD_adj-sigDEG_BD_means_adj
```

```{r, echo=FALSE, fig.width=9,fig.cap="**Figure 15.** A heatmap of the significantly differntially expressed genes (adjusted p-value where alpha = 0.1) that correspond to mitochondrial membrane GO terms generated from the calculated differences from the means. Orange represents genes that were upregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata* hosting Symbiodiniaceae strain B01 and blue represents genes that were downregulated in *O.faveolata* hosting strain D compared to *O.faveolata* hosting strain B01."}
heatmap_sigDEG_means_adj = heatmap.2(as.matrix(explc_sig_adj), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG Mean, padj < 0.1",
                           margin = c(5,15))
```


### Significant (p-value < 0.01) DEG - Mitochondria

Now that the list has been filtered to only significantly differentially expressed genes with a p-value at alpha = 0.01, all genes that correspond to the GO terms of mitochondrial membrane function can be selected for and stored in `sigDEG_BD_01`. The number of significantly differentially expressed genes that are associated with the mitochondrial membrane function terms are counted and reported below. 
```{r, warning=FALSE, message=FALSE}
sigDEG_BD_01 = iso2go %>%
  left_join(rlog_BD_01) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
nrow(sigDEG_BD_01)
```

Here there are only 41 of the 68 DEGs that had annotated GO terms. So a heatmap with these raw values can be generated. 

```{r, echo=FALSE, fig.width=9, fig.height=10, fig.cap="**Figure 16.** A heatmap of the significantly differntially expressed genes (p-value where alpha = 0.01) that correspond to mitochondrial membrane GO terms. Orange represents genes that were upregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata* hosting Symbiodiniaceae strain B01 and blue represents genes that were downregulated in *O.faveolata* hosting strain D compared to *O.faveolata* hosting strain B01 "}
heatmap_sigDEG_01 = heatmap.2(as.matrix(sigDEG_BD_01), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG, pval < 0.01",
                           margin = c(5,15))
```

Noting that many of the terms are un-annotated, a heatmap can be generated by calculating the difference from the means of the adjusted p-values of the differentially expressed genes. 

```{r}
sigDEG_BD_means_01 = apply(sigDEG_BD_01, 1, mean)
explc_sig_01 = sigDEG_BD_01-sigDEG_BD_means_01
```

```{r, echo=FALSE, fig.width=9, fig.height=10, fig.cap="**Figure 17**. A heatmap of the significantly differntially expressed genes (p-value where alpha = 0.01) that correspond to mitochondrial membrane GO terms generated from the calculated differences from the means. Orange represents genes that were upregulated in *Orbicella faveolata* hosting Symbiodiniaceae strain D compared to *O.faveolata* hosting Symbiodiniaceae strain B01 and blue represents genes that were downregulated in *O.faveolata* hosting strain D compared to *O.faveolata* hosting strain B01." }
heatmap_sigDEG_means_01 = heatmap.2(as.matrix(explc_sig_01), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG Mean, pval < 0.01",
                           margin = c(5,15))
```


```{r, echo=FALSE, eval=FALSE}
#Make this list of genes and expression values into a csv file for future reference
#Also attach the jpg image of the heatmap so that all the gene symbols are visible
write.csv(explc_sig_adj, "rare_0.1adjDEG_BD_genesymbols.csv", quote = F)
write.csv(explc_sig_01, "rare_0.01pvalDEG_BD_genesymbols.csv", quote = F)
knitr::opts_chunk$set(echo = TRUE)
knitr::include_graphics("Ofav Adj 0.1 Sig DEG Mean Heatmap.jpg")
```


## Conclusions 

In conclusion, by rarefying our count data, we were able to perform a more normalized analysis of differentially expressed genes across our samples with our results less confounded by coverage variability. However, it should be noted that the process of rarefying selects a random subset of the data, thus some data, most likely representative of weaker - though significant - associations may be lost. In our analysis of GO terms enriched in coral hosting strain D compared to strain B01, we identified an upregulation of mitochondrial membrane functions, including genes involved in the regulation of reactive oxidative species. These data were supported by genome-wide analysis of differentially expressed genes, regardless of their GO-term association, which, together, identified an upregulation in genes such as Cytochrome C Oxidase (major oxygen consumer in the electron transport chain of cellular respiration), as well as genes involved in peroxide metabolism. The upregulation of these genes under basal conditions indicates that the cell may be more equipped to handle, and quench, oxidative stress. At least one gene involved in the heat stress response, HSP110, was downregulated in coral hosting strain D compared to strain B01. Additionally, several innate immune response genes were identified as upregulated in coral hosting strain D compared to strain B01, a result which agrees with previous studies suggesting the coral may be responding to the “non-preferred” symbiont as a pathogen instead of as an entirely beneficial symbiont. It is important to note that there appears to be a reallocation of resources in coral in symbiosis with strain D, as there is a downregulation of GO terms associated with cell migration/adhesion when compared to those in symbiosis, indicating that the coral host may be growing at a slower rate when hosting the “non-preferred” symbiont. These data agree with previous research that has implicated a fitness tradeoff in corals hosting distinct symbiont strains. While coral hosting the “non-preferred” symbiont may be more tolerant of oxidative stressors that arise in the environment, they may have lower fitness during non-stressful conditions.