---
title: "Orbicella Faveolata Reuptake"
author: "Maria Ingersoll, Jamie Poirer, Susritha Kopparapu"
date: "4/29/2021"
output:
  html_document: default
---


## Introduction

## Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### R Version
R version 4.0.2 was used for this analysis. 

#### Packages
The following packages were used to clean, analyze, and visualize data. Packages were installed with [Bioconductor](https://bioconductor.org/biocLite.R) version 3.12 or the [Cran Repository](http://cran.us.r-project.org).   
```{r, message=FALSE, warning=FALSE}
library("DESeq2")                   #Version 1.30.1
library("arrayQualityMetrics")      #Version 3.46.0
library("dplyr")                    #Version 1.0.5
library("RColorBrewer")             #Version 1.1.2
library("gplots")                   #Version 3.1.1
library("tidyverse")                #Version 1.3.0
library("pheatmap")                 #Version 1.0.12
library("vegan")                    #Version 2.5.7
library("ggplot2")                  #Version 3.3.3
library("ggrepel")                  #Version 0.9.1

#--------------
library("affycoretools")            #Version 1.60.1
library("genefilter")               #Version 1.70.0
```

```{r, eval=FALSE, echo=FALSE}
packageVersion("DESeq2")
packageVersion("arrayQualityMetrics")
packageVersion("dplyr")
packageVersion("RColorBrewer")
packageVersion("gplots")
packageVersion("tidyverse")
packageVersion("pheatmap")
packageVersion("vegan")
packageVersion("ggplot2")
packageVersion("ggrepel")


#packageVersion("affycoretools")
#packageVersion("genefilter")

```

## Raw Data Filtering

```{r,echo=FALSE, eval=FALSE}
setwd("/Users/mariaingersoll/Desktop/BU2020-2021/Ecol_and_Enviro_Genomics/BI586-git/Ofav_uptake_finalproj")
```

```{r, echo=FALSE}
setwd("~/BI 586/FinalProject")
```


Counts data was stored in a text file that was read in as a data frame to a variable called `countData`. Note that this data was pre-filtered for a base mean of 3. The column names are renamed based on the the strain type and replicate number for readability sake. The length of the data frame is reported and it represents the number of isoforms.
```{r}
countData <- read.table("orbfav_uptake_counts.txt")
newColNames = c("BR.A", "BR.B", "BR.C", "BR.D", "CR.A", "CR.B", "CR.C","DR.A","DR.B", "DR.C", "DR.D")
colnames(countData)=paste(newColNames)
length(countData[,1])
```

The total sum of isoform counts for each condition are displayed below in a barplot. The minimum and maximum are as reported below. 
```{r, echo=FALSE, fig.cap="**Figure 1.** All raw sequencing count values from each Orbicella faveolata sample demonstrating various degrees of library sizes. Green bars are O. faveolata hosting Symbiodiniaceae strain B01 (replicates: BR.A, BR.B, BR.C, BR.D). Red bars are O. faveolata hosting Symbiodiniaceae strain C (triplicates: CR.A,CR.B, CR.c). Blue bars are O. faveolata hosting Symbiodiniaceae strain D (replicates: DR.A, DR.B, DR.C, DR.D).", fig.width=10}
totalCounts=colSums(countData)
barplot(totalCounts, col=c("green",  "green", "green", "green", "red", "red", "red", "blue", "blue" , "blue" , "blue"), ylab="Raw Counts", xlab="Coral Hosting Symbiodiniaceae Strains")
```

```{r}
min(totalCounts)
max(totalCounts)
```


Due to the small library sizes of one replicate from the strain B, one replicate from the strain D conditions, and all replicates from O. faveolata hosting Symbiodiniaceae strain C were removed. `newCounts` is generated from the original `countData` but only selecting the samples that have larger library size and excluding the following ones as they are too small. 
```{r}
newCounts = countData %>% 
  dplyr::select(-BR.D, -CR.A, -CR.B, -CR.C, -DR.B)
```


```{r, echo=FALSE, fig.cap="**Figure 2.** Selected raw sequencing count values from each Orbicella faveolata sample demonstrating various degrees of library sizes. Green bars are O. faveolata hosting Symbiodiniaceae strain B01 (triplicates: BR.A, BR.B, BR.C). Blue bars are O. faveolata hosting Symbiodiniaceae strain D (triplicates: DR.A, DR.C, DR.D). One replicate from the strain B condition, one replicate from the strain D condition, and all replicates from O. faveolata hosting Symbiodiniaceae strain C were removed prior to this analysis due to small library sizes."}
totalCounts=colSums(newCounts)
barplot(totalCounts, col=c("green", "green", "green", "blue" , "blue" , "blue"), ylab="Raw Counts", xlab="Coral Hosting Symbiodiniaceae Strains")
```

Below the minimum and maximum summed counts, after removal of small library sizes, are reported. 
```{r}
min(totalCounts)
max(totalCounts)
```

*Note: In the above barplot, we see that despite the previous filtering, there are still conditions that have large library sizes (BR.C, DR.A, and DR.C) compared to other conditions that have very low library sizes (BR.A, BR.B, annd DR.B). This inconsistency in library size must be kept in mind for conclusions drawn from later analysis of the differential expression of genes between coral hosting strain B and coral hosting strain D.*

- Rareify counts, cut down the high counts to 100,000
- First look at all the counts data
```{r, eval=FALSE}
# rarecurve(newCounts,step=100,label=FALSE)
t_newCounts = t(newCounts)
rarefy(t_newCounts, 100000, se=FALSE, MARGIN=1)
head(t_newCounts)
unt_newCounts = t(t_newCounts)
rare_totalCounts=colSums(unt_newCounts)
head(rare_totalCounts)

rrarefy(t_newCounts, 100000)
unt_newCounts = t(t_newCounts)
rare_totalCounts=colSums(unt_newCounts)
head(rare_totalCounts)

rarecurve(t_newCounts,step=100000)
total <- rowSums(t_newCounts)
subset = subset(total, total >50000)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.12")
BiocManager::install("metaseqR2")
library(metaseqR2)
downsample.counts(subset, seed=42)
```

## Outlier Analysis

For outlier analysis, we first create a character vector holding the names of the strains in each sample. The vector is converted to a dataframe and is renamed `colData` for further use. 
```{r, eval=FALSE}
treat=c( "B", "B", "B", "D", "D", "D")
g=data.frame(treat)
colData<- g
colData
```

`DESeqDataSetFromMatrix` was run on the `newCounts` which represents the raw counts of the selected samples. The column data is set to the strain type and the design to the treatment vector so that it may dictate how the counts for each gene depend on the variables in colData. The `DESeq` step was run on the results of the prior step.  
```{r results="hide", message=FALSE, eval=FALSE}
dds<-DESeqDataSetFromMatrix(countData=newCounts, colData=colData, design=~treat)
dds<-DESeq(dds)
```

The `vst` function is the variance-stabilizing transformation which models the mean-variance relationship of replicates. The assay of dds generates a table with the mean-variance of each of the isoform treatments. The as.data.frame(colData(rl))) generates a table with the columns and a sizeFactor value that is converted into the formal class AnnotatedDataFrame. Both values are passed are used to create an ExpressionSet object which is used to generate the array’s quality metrics using the treat variable as the intgroup.
```{r, eval=FALSE}
vsd.ge=assay(vst(dds))
rl=vst(dds)
e=ExpressionSet(assay(rl), AnnotatedDataFrame(as.data.frame(colData(rl))))
v = setwd("~/BI 586/FinalProject")
arrayQualityMetrics(e,outdir=v,intgroup=c("treat"),force=T)
```
The array’s quality metrics showed that there was only one outlier (for the second replicate of the O.fav with Symbiodiniaceae strain B) by only one detection method (the MA plot). The the other two detection methods: by distances between arrays and by boxplots showed no outliers in any of the samples.


`results` extracts the results table from the DESeq analysis and the DESeqDataSet is used to generate the dispersion plot.
```{r, eval=FALSE}
head(dds)
res<- results(dds)
```
- The above code gives the log2 fold change (MLE): treat D vs B
- this means that pos log2FC values indicate D is upregulated compared to B
- The name provided in the second element (B) is the baseline

Spread of data is shown with the dispersion plot below. 
```{r, echo=FALSE, fig.cap="Figure 3. Scatter plot of dispersion estimates per gene (on the y-axis) versus the mean of normalized counts (on the x-axis) of isoforms identified in the treatment samples of Orbicella faveolata. Gene-wise dispersion estimates are in black, fitted estimates are in red, and final estimates are in blue as indicated by the legend.", eval=FALSE}
plotDispEsts(dds, main="Dispersion plot Uptake")
```

- Now we are going to retrieve the rlog data, which will give us better transformation when size factors vary across samples
- regularized log transformation, transforms count data to log2 scale to minimize differences between samples for rows with small counts

```{r, eval=FALSE}
rld <- rlogTransformation(dds, blind=TRUE)
head(assay(rld))

```

```{r, echo=FALSE, fig.cap="Figure 3. Histogram of r-log transformations", eval=FALSE}
hist(assay(rld))
```


- The assay function allows you to access the matrix-like data, so that you can view it bc head(rld) gives you just a summary
- The histogram above displays the frequency of each binned rld count value


- Making a sample distance heatmap
- as.matrix attempts to turn its argument into a matrix
- dist computes and returns the distance matrix computed using the specified distance measure to compute distances between the rows of a data matrix (assay(rld))
- Create a sample heatmap of the relatedness based on the count distance/difference between samples
```{r results="hide", message=FALSE, eval=FALSE}
sampleDists <- as.matrix(dist(t(assay(rld))))
```

```{r, echo=FALSE, fig.cap="Figure 5. Sample distance heatmap where BR.C is clustering more with D.", eval=FALSE}
heatmap.2(as.matrix(sampleDists), key=F, trace="none",
          col=colorpanel(100, "black", "white"),
          margin=c(10, 10))
```

- Now look at results of dds comparing (average) B and D
- The second term is the "control" we will use B as control in this case

```{r, eval=FALSE}
resBD <- results(dds, contrast=c("treat","D","B"))
```
- How many gene counts have FDR < 10%
- Analyze the count differences between B and D
- This is all quality control
```{r, eval=FALSE}
table(resBD$padj<0.1)
summary(resBD)
```

```{r, echo=FALSE, fig.cap="Figure 6. MA plot between B and D.", eval=FALSE}
plotMA(resBD, main="B vs D")
```



- Now make a table of the differential expression values (log2fc) between B and D
```{r, eval=FALSE}
write.table(resBD, file="BD_DE.txt", quote=F, sep="\t")
head(read.delim("BD_DE.txt"))
```
- "BD_DE.txt" contains the log2FC, the pvalue, and the padj, as well as some other data that we don't really need

- Get the p-values from your data file
- Create valBD that is a table from resBD of just the pvalues and adjusted p-values (p-adj)
- Then give the columns approprite names
```{r, eval=FALSE}
valBD=cbind(resBD$pvalue, resBD$padj)
head(valBD)
colnames(valBD)=c("pval.BD", "padj.BD")
length(valBD[,1])
```
- Return a logical vector in which both cases (pval and padj) are complete (no missing values)
```{r, eval=FALSE}
table(complete.cases(valBD))
```
- Make a table (that you can visualize via assay) of the rlogdata and pvals
- Then make the column names for rld the same ones that were in colData$treat
```{r, eval=FALSE}
rlog=rlogTransformation(dds, blind=TRUE) 
rld=assay(rlog)
head(rld)
colnames(rld)=paste(colData$treat)
head(rld)
length(rld[,1])
```
- Combine rld with your table of pval and padj values (valBD)
```{r, eval=FALSE}
rldpvals=cbind(rld,valBD)
head(rldpvals)
dim(rldpvals)
table(complete.cases(rldpvals))
```
- Create a new csv file from rldpvals
```{r, eval=FALSE}
write.csv(rldpvals, "BI586_uptake_RLDandPVALS.csv", quote=F)
```
- Now we will create a heatmap of sample distances based on pvalues
- Make a new variable for your csv file
```{r, eval=FALSE}
rldpvals <- read.csv(file="BI586_uptake_RLDandPVALS.csv", row.names=1)
head(rldpvals)
```
- Make rld that a variable that just returns the first six columns with values (rownames then columns 1-6)
```{r, eval=FALSE}
rld=rldpvals[,1:6]
head(rld)
```
- dist computes and returns the distance matrix computed using the specified distance measure to compute distances between the rows of a data matrix (rld); determine the overall differences in expression between each sample based on the rlog values
- Turn the distances into a matrix
```{r,eval=FALSE}
sampleDists <- dist(t(rld))
sampleDistMatrix <- as.matrix(sampleDists)
```
- Create a vector of treatment titles for the matrix and apply them to the rows and columns
```{r, eval=FALSE}
treat=c("B", "B", "B", "D", "D", "D")
colnames(sampleDistMatrix)=paste(treat)
rownames(sampleDistMatrix)=paste(treat)
```
- Install some packages to make heatmaps
```{r results="hide", eval=FALSE}
heat.colors = colorRampPalette(rev(c("blue","yellow")),bias=0.3)(100)
quartz()
```
- Apply your matrix of overall expression distance to a heatmap
```{r, echo=FALSE, fig.cap="Figure 7. Sample Distance Matrix between B and D.", eval=FALSE}
pheatmap(sampleDistMatrix,color = heat.colors,cex=0.9,border_color=NA,cluster_rows=T,cluster_cols=T)
```


- Now going to perform PCA on the data to visualize overall effect of experimental covariates and batch effects
- t transposes the rld table and makes all the rows columns, and the columns rows. So here the isoform names that were rownames are now the column names. 
```{r,eval=FALSE}
rld_t=t(rld)
head(rld)
```


```{r results="hide",eval=FALSE}
head(rld_t)
```
- prcomp performs a pca on given data matrix and returns results as an object of class prcomp
```{r results="hide", warning=FALSE, eval=FALSE}
pca <- prcomp(rld_t,center = TRUE, na.action=na.omit)
head(pca)
```

- sdev is the standard deviations of the principal components
- Using the sdev, calculate the proportion that each PC corresponds to the variance
- Then round PC1 and PC2 (times 100, to 1 sigfig)
```{r,eval=FALSE}
li <- pca$sdev^2 / sum(pca$sdev^2)
pc1v <- round(li[1] * 100, 1)
pc2v <- round(li[2] * 100, 1)
```
- x from prcomp seems like it's the coordinates of each treatment in the PCA
- turn pca$x into a dataframe
```{r,eval=FALSE}
pca_s <- as.data.frame(pca$x)
head(pca_s)
```
- Just take PC1 and PC2 and add your sample and treatment names
```{r,eval=FALSE}
pca_s <- pca_s[,c(1,2)]
head(pca_s)
pca_s$Samples = row.names(pca_s)
head(pca_s)
pca_s$treat=colData$treat
head(pca_s)
```
- Creating your PCA plot
- aes is aesthetic mappings, how variables are mapped to visual properties
```{r, echo=FALSE, fig.cap="Figure 8. Principal Component Analysis of B vs D.", eval=FALSE}
cbPalette <- c("darkorchid4","firebrick4")
ggplot(pca_s, aes(PC1, PC2, color = treat, pch = treat)) +
  geom_point(size=3) +
  #  geom_text_repel(aes(label=Samples)) +
  scale_colour_manual(values=cbPalette)+
  theme_bw() +
  # geom_density2d(alpha=.5)+
  geom_polygon(alpha=.2)+
  xlab(paste0("PC1: ",pc1v,"% variance")) +
  ylab(paste0("PC2: ",pc2v,"% variance")) 
```

- You can see from this plot that the B samples have very high variance and that there's no overlap between the samples

- Perform adonis to determine whether the conditions are significantly different
```{r,eval=FALSE}
adonis(pca$x ~ treat, data = pca_s, method='eu', na.rm = TRUE)
```

Permutation: free
Number of permutations: 719

Terms added sequentially (first to last)

          Df SumsOfSqs MeanSqs F.Model      R2 Pr(>F)
treat      1    6375.9  6375.9  1.2256 0.23454    0.2
Residuals  4   20809.1  5202.3         0.76546       
Total      5   27185.0                 1.00000  


```{r,eval=FALSE}
colData$D<-factor(colData$treat, levels=c("D","B"))

resD <- results(dds, contrast=c("treat","D","B"))
#how many FDR < 10%?
table(resD$padj<0.01)
#FALSE  TRUE 
# 3294     4 
summary(resD)
  #out of 16148 with nonzero total read count
  #adjusted p-value < 0.1
  #LFC > 0 (up)       : 2, 0.012%
  #LFC < 0 (down)     : 27, 0.17%
  #outliers [1]       : 19, 0.12%
  #low counts [2]     : 12831, 79%
  #(mean count < 10)

nrow(resD[resD$padj<0.05 & !is.na(resD$padj),])  # Num significantly differentially expressed genes excluding the no/low count genes   #20

dev.off()
plotMA(resD, main="B vs D")
plotMA(resD, main="B vs D", ylim=c(-2,2))

results <- as.data.frame(resD)
head(results)

nrow(resD[resD$padj<0.1 & resD$log2FoldChange > 0 & !is.na(resD$padj),])

nrow(resD[resD$padj<0.1 & resD$log2FoldChange < 0 & !is.na(resD$padj),])
#UP 2
#DOWN 27

write.table(resD, file="D_2021.txt", quote=F, sep="\t")
cd <- read.table("D_2021.txt")
head(cd)
```

- Create a term go_input_D that has the -logpval as a neg number if the l2fc is < 0 and a pos number if the l2fc is > 0
- This term will go into the go analysis
```{r,eval=FALSE}
cd
go_input_D = cd %>%
  tibble::rownames_to_column(var = "iso") %>%
  mutate(mutated_p = -log(pvalue)) %>%
  mutate(mutated_p_updown = ifelse(log2FoldChange < 0, mutated_p*-1, mutated_p*1)) %>%
  na.omit() %>%
  select(iso, mutated_p_updown)
head(go_input_D)
colnames(go_input_D) <- c("gene", "pval")
head(go_input_D)
write.csv(go_input_D, file="D_GO.csv", quote=F, row.names=FALSE)
```


ACTUAL DIFFERENTIAL GENE EXPRESSION ANALYSIS STARTS HERE
- This is just the preliminary GO enrichment analysis of all the genes. The way gomwu works is that it performs rankings and gives you enrichment of GO terms in the ranked list. So GO terms that show up may not actually contain only genes that are significantly differentially expressed, that gene set is just enriched

- Plotting genes within each GO category between B and D
- MF: Molecular function
- BP: Biological process
- CC: Cellular component

- Manipulated BD_DE.txt in excel following Sarah's verbal instructions and loaded it here
- BD_DE_GO has the mutated -log(p-values) \
- From the BI586_uptake file, we got log2 fold change (MLE): treat D vs B
- Reminder: this means that pos log2FC values indicate D is upregulated compared to B; The name provided in the second element (B) is the baseline

```{r,eval=FALSE}
BD_DE_GO = read.csv("D_GO.csv")
```
- Creating dendrograms (more info can be found in the r file GO_MWU.R)
- First, run MF; edit the input stuff to match what you're running
- BD_DE_GO.csv has two columns of comma-separated values: gene id, continuous measure of significance. To perform standard GO enrichment analysis based on Fisher's exact test, use binary measure (0 or 1, i.e., either significant or not).
- orb_fav_iso2go.txt is two-column, tab-delimited, one line per gene, multiple GO terms separated by semicolon. If you have multiple lines per gene, use nrify_GOtable.pl prior to running this script.
- go.obo is downloaded from http://www.geneontology.org/GO.downloads.ontology.shtml
- goDivision is either MF, BP, or CC

CONTROL REPLACE ofav to isogroup in D_GO.CSV so they can be linked to the go annotations file
```{r,eval=FALSE}
head(read.csv("D_GO.csv"))
input="D_GO.csv"
goAnnotations="orb_fav_iso2go_new.tab"
goDatabase="go.obo"
goDivision="MF"
source("gomwu.functions.R")
```

- Calculating stats. It might take ~3 min for MF and BP. Do not rerun it if you just want to replot the data with different cutoffs, go straight to gomwuPlot. If you change any of the numeric values below, delete the files that were generated in previous runs first.
- perlPath: replace with full path to perl executable if it is not in your system's PATH already
- largest: a GO category will not be considered if it contains more than this fraction of the total number of genes
- smallest: a GO category should contain at least this many genes to be considered
- clusterCutHeight: threshold for merginng similar (gene-sharing) terms. See gomwu README for details
```{r,eval=FALSE}
gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25)
```
NEW TRANSCRIPTOME: 18  GO terms at 10% FDR

- Old transcriptome: Returned 9 GO terms at 10% FDR
- Plot results in a dendrogram
- absValue=-log(0.05,10); genes with the measure value exceeding this will be counted as "good genes". Specify absValue=0.001 if you are doing Fisher's exact test for standard GO enrichment or analyzing a WGCNA module (all non-zero genes = "good genes").
- level1=0.1: FDR threshold for plotting. Specify level1=1 to plot all GO categories containing genes exceeding the absValue.
- level2=0.05: FDR cutoff to print in regular (not italic) font.
- level3=0.01: FDR cutoff to print in large bold font.
- txtsize=1.5: decrease to fit more on one page, or increase (after rescaling the plot so the tree fits the text) for better "word cloud" effect
- treeHeight=0.5: height of the hierarchical clustering tree
- colors=c("dodgerblue2","firebrick1","skyblue","lightcoral") # these are default colors, un-remark and change if needed
- Typing resultsMF gives you the data in a written format
```{r, warning=FALSE, eval=FALSE}
quartz()
resultsMF=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.5,
                  treeHeight=0.5,
)
resultsMF
```

- Doesn't look that pretty, but if you really want it to look nice, run the above code in base R and then you can manipulate the size of the image to your liking, or export it as a pdf and fiddle around with the size
- Red is upregulated in D and blue is downregulated in D (upregulated in B)
- The GO term we are most interested in here is Oxidoreductase, remember this for later

- Now let's run the above protocol for BP
```{r,eval=FALSE}
head(read.csv("D_GO.csv"))
input="D_GO.csv"
goAnnotations="orb_fav_iso2go_new.tab"
goDatabase="go.obo"
goDivision="BP"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25)
```
NEW TRANSCRIPTOME: 10 GO Terms at 10% FDR 

```{r, warning=FALSE, eval=FALSE}
quartz()
resultsBP=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.5,
                  treeHeight=0.5,
)
resultsBP
```

- Old transcriptome: Returned 6 GO terms at 10% FDR
- Here we are interested in the positive regulation of cell proliferation

- Now let's run it for CC
```{r,eval=FALSE}
head(read.csv("D_GO.csv"))
input="D_GO.csv"
goAnnotations="orb_fav_iso2go_new.tab"
goDatabase="go.obo"
goDivision="CC"
source("gomwu.functions.R")

gomwuStats(input, goDatabase, goAnnotations, goDivision,
           perlPath="perl",
           largest=0.1,
           smallest=5,
           clusterCutHeight=0.25)
```
NEW TRANSCRIPTOME: 23  GO terms at 10% FDR


```{r, warning=FALSE, eval=FALSE}
quartz()
resultsCC=gomwuPlot(input,goAnnotations,goDivision,
                  absValue=1,
                  level1=0.1,
                  level2=0.05,
                  level3=0.01,
                  txtsize=1.5,
                  treeHeight=0.5,
)
resultsCC
```

- Old transcriptome: Returned 18 GO terms of interest
- Interested in the mitochondrial complexes



INVESTIGATING DIFFERENTIAL EXPRESSION OF GENES IN SPECIFIC GO CATEGORIES
- Using Dan Wuitchik's code: https://github.com/wuitchik/Divergent-thermal-challenges-elicit-convergent-stress-signatures-in-aposymbiotic-Astrangia-poculata/blob/master/Astrangia_HotCold.pdf
- First, looking at positive regulation of cell proliferation (from BP, GO:0008284)
- Load packages
- Set up the variable iso2go which contains the ofav gene id and all of its associated GO terms


```{r, eval=FALSE}
iso2go = read.table("orb_fav_iso2go_new.tab", sep="\t")
head(iso2go)
colnames(iso2go) <- c("gene_ID", "GO.terms")
```
- Obtain the gene symbol from the gene descriptions in the file orb_fav_iso2gene.tab
- Create the variable gene which contains the gene id, the full description, and the gene symbol
```{r,eval = FALSE}
newnames = c("gene_ID", "description" )
iso2gene = read.table("orb_fav_iso2gene_new.tab", sep="\t")
colnames(iso2gene) <- newnames

gene = iso2gene %>%
       mutate(gene_symbol = gsub(".* GN=", "", description)) %>%
       mutate(gene_symbol = gsub(" .*", "", gene_symbol))
       
head(gene)
```
- To get the rlog foldchange values, I'm going to manipulate the BI586 uptake file (which I will save in a variable called BD_DE)
```{r,eval=FALSE}
head(read.csv("BI586_uptake_RLDandPVALS.csv"))
BD_DE = read.csv("BI586_uptake_RLDandPVALS.csv")
library(dplyr)
head(BD_DE)
rlog_BD = BD_DE[,1:7] %>%
  dplyr::rename(gene_ID = X)
head(rlog_BD)
```
- rlog_BD just has the isoform names (called gene_ID) and the log2FC for each sample
- could have also used the tidyverse select function instead of [,1:7] here, Dan says it's better to keep the same format so using select would have been better

- Next, make a variable that contains all the genes that correspond to the GO term for pos reg of cell prolif GO:0008284 
```{r, warning=FALSE, eval=FALSE}
GO_0008284 = iso2go %>%
  filter(str_detect(GO.terms, "GO:0008284")) %>%
  left_join(rlog_BD) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
```

```{r, eval=FALSE}
head(GO_0008284)
colnames(rlog_BD)
```

- Plot a heatmap of the differential expression of ALL the genes that correspond to GO_0008284 pos reg of cell prolif
```{r,eval=FALSE}
heatmap_GO_0008284 = heatmap.2(as.matrix(GO_0008284), Rowv = TRUE, Colv = FALSE, scale = "row",
          dendrogram = "both",
          trace = "none",
          main = "GO_0008284 positive regulation of cell proliferation",
          margin = c(5,15))
```

- Yellow to white correspond to upregulation and red corresponds to downregulation

- Now doing the same for oxidoreductase
- Oxidoreductase complex from CC (GO:1990204); oxidoreductase activity from MF (GO:0016491)
- To filter for multiple terms, use | (the symbol that is shift back slash) and quotes only outside all the terms
```{r,eval=FALSE}
oxidoreductase = iso2go %>%
  filter(str_detect(GO.terms, "GO:1990204|GO:0016491")) %>%
  left_join(rlog_BD) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
head(oxidoreductase)
nrow(oxidoreductase)
```
- 92 genes that correspond to these two GO terms
Plot a heatmap of the differential expression of ALL the genes that correspond to the oxidoreductase GO terms above
```{r,eval=FALSE}
heatmap_oxidoreductase = heatmap.2(as.matrix(oxidoreductase), Rowv = TRUE, Colv = TRUE, scale = "row",
                               dendrogram = "both",
                               trace = "none",
                               main = "GO Oxidoreductase",
                               margin = c(5,15))
```

- Notice that in the above heatmaps, we clustered the columns (the samples) as well as the rows. When we clustered the columns in the oxidoreductase heatmap, B.2 clusters with the Ds. This may be an affect of the count values we obtained from each sample, as B.2 was the B sample with the high count values.
- One way to mitigate this is to take the means of the rows, subtract the mean from the original value, and plot the result instead of the original value
- We'll also be filtering for only those genes that are significantly differentially expressed between recruits hosting D vs B

- Now we want to make new heatmaps of just those genes in cell prolif and oxidoreductase that are significantly differentially expressed (we will use an alpha of 0.1)

- Filter the BI586 uptake data for all genes significantly differentially expressed (alpha = 0.1 for this section, we will use the more stringent 0.05 later)
```{r,eval=FALSE}
BD_DE_10 = filter(BD_DE, pval.BD <= 0.1, preserve = TRUE)
head(BD_DE_10) 
rlog_BD_10 = BD_DE_10[,1:7] %>%
  dplyr::rename(gene_ID = X)
head(rlog_BD_10)
nrow(rlog_BD_10)
```
- rlog_BD_10 has all the genes that are sig DE with alpha=0.1 and does not contain the p-val nor the adj p-val

- Now search within this list for all genes that correspond to the GO term for pos reg of cell prolif and call this new list GO_0008284_sig
```{r,eval=FALSE}
GO_0008284_sig = iso2go %>%
  filter(str_detect(GO.terms, "GO:0008284")) %>%
  left_join(rlog_BD_10) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
GO_0008284_sig
nrow(GO_0008284_sig)
```
- We have 6 genes, plot them in a heatmap but this time use the calculated difference from the mean
```{r,eval=FALSE}
GO_0008284_means_sig = apply(GO_0008284_sig, 1, mean)
explc_cellprolif = GO_0008284_sig-GO_0008284_means_sig
head(explc_cellprolif)
col0=colorRampPalette(rev(c("chocolate1","#FEE090","grey10", "cyan3","cyan")))(100)
heatmap_GO_0008284_sig = heatmap.2(as.matrix(explc_cellprolif), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                               dendrogram = "both",
                               trace = "none",
                               main = "GO_0008284 positive regulation of cell proliferation",
                               margin = c(5,15))
```

- In this heatmap, I've used a new color palette. Now blue is downregulated and orange is upregulated

- Now, search within rlog_BD_10 for all the sig DE genes that correspond to the oxidoreductase terms
```{r,eval=FALSE}
oxidoreductase_sig = iso2go %>%
  filter(str_detect(GO.terms, "GO:1990204|GO:0016491")) %>%
  left_join(rlog_BD_10) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
oxidoreductase_sig
```
- We have 4 genes here, plot the difference from the mean on a heatmap
```{r,eval=FALSE}
oxidoreductase_means_sig = apply(oxidoreductase_sig, 1, mean)
explc_oxr = oxidoreductase_sig-oxidoreductase_means_sig
head(explc_oxr)

heatmap_oxidoreductase_sig = heatmap.2(as.matrix(explc_oxr), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                                   dendrogram = "both",
                                   trace = "none",
                                   main = "GO Oxidoreductase",
                                   margin = c(5,15))
```

- For the final step, I'm going to make a heatmap of all the genes that are differentially regulated (alpha = 0.05 and alpha = 0.01 to account for the count differences) that will allow for further exploration

```{r,eval=FALSE}
BD_DE_05 = filter(BD_DE, pval.BD <= 0.05, preserve = TRUE)
head(BD_DE_05) 
rlog_BD_05 = BD_DE_05[,1:7] %>%
  dplyr::rename(gene_ID = X)
head(rlog_BD_05)
nrow(rlog_BD_05)

BD_DE_01 = filter(BD_DE, pval.BD <= 0.01, preserve = TRUE)
head(BD_DE_01) 
rlog_BD_01 = BD_DE_01[,1:7] %>%
  dplyr::rename(gene_ID = X)
head(rlog_BD_01)
nrow(rlog_BD_01)
```
- alpha of 0.05 yields 400 terms which is unmanageable
- alpha of 0.01 yields 120 terms which is better
- Let's also try padj alpha of 0.1:
```{r,eval=FALSE}
BD_DE_adj = filter(BD_DE, padj.BD <= 0.1, preserve = TRUE)
head(BD_DE_adj) 
rlog_BD_adj = BD_DE_adj[,1:7] %>%
  dplyr::rename(gene_ID = X)
head(rlog_BD_adj)
nrow(rlog_BD_adj)
```
- padj yields 29 terms which is pretty manageable

```{r,eval=FALSE}
sigDEG_BD_adj = iso2go %>%
  left_join(rlog_BD_adj) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
head(sigDEG_BD_adj)
nrow(sigDEG_BD_adj)

sigDEG_BD_01 = iso2go %>%
  left_join(rlog_BD_01) %>%
  left_join(gene) %>%
  mutate(gene_symbol = make.names(gene_symbol, unique = TRUE)) %>%
  column_to_rownames(var = "gene_symbol") %>%
  dplyr::select(-GO.terms, -description, -gene_ID) %>%
  drop_na() %>%
  dplyr::select(sort(current_vars()))
head(sigDEG_BD_01)
nrow(sigDEG_BD_01)
```
- sigDEG_BD_adj are all my significant DEGs between B and D in the right format with padj of 0.1
- Only 17 out of the 29 DEGs had annotated GO terms
- sigDEG_BD_01 are all my significant DEGs between B and D in the right format with pval of 0.01
- Only 74 out of the 120 DEGs had annotated GO terms
- First making a heatmap of the raw values for the adj
```{r,eval=FALSE}
heatmap_sigDEG_adj = heatmap.2(as.matrix(sigDEG_BD_adj), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG, padj < 0.1",
                           margin = c(5,15))
```
- A lot of the terms are un-annotated
- Now calculating the difference from the mean of the padj DEGs and making that a heatmap
```{r,eval=FALSE}
sigDEG_BD_means_adj = apply(sigDEG_BD_adj, 1, mean)
explc_sig_adj = sigDEG_BD_adj-sigDEG_BD_means_adj
heatmap_sigDEG_means_adj = heatmap.2(as.matrix(explc_sig_adj), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG Mean, padj < 0.1",
                           margin = c(5,15))
```
- Do the same with the pval of 0.01
```{r, eval=FALSE}
heatmap_sigDEG_01 = heatmap.2(as.matrix(sigDEG_BD_01), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG, pval < 0.01",
                           margin = c(5,15))

sigDEG_BD_means_01 = apply(sigDEG_BD_01, 1, mean)
explc_sig_01 = sigDEG_BD_01-sigDEG_BD_means_01
heatmap_sigDEG_means_01 = heatmap.2(as.matrix(explc_sig_01), col = col0, Rowv = TRUE, Colv = TRUE, scale = "row",
                           dendrogram = "both",
                           trace = "none",
                           main = "Significant DEG Mean, pval < 0.01",
                           margin = c(5,15))
```


- Make this list of genes and expression values into a csv file for future reference
- Also attach the jpg image of the heatmap so that all the gene symbols are visible
```{r, eval=FALSE}
write.csv(explc_sig_adj, "0.1adjDEG_BD_genesymbols.csv", quote = F)
write.csv(explc_sig_01, "0.01pvalDEG_BD_genesymbols.csv", quote = F)
knitr::opts_chunk$set(echo = TRUE)
knitr::include_graphics("Ofav Adj 0.1 Sig DEG Mean Heatmap.jpg")
```


- MAIN FINDINGS:
- From GO MWU analysis of ranked GO term enrichment, we found that Postitive Regulation of Cell Proliferation and various Oxidoreductase activities are enriched in samples hosting D compared to B recruits
- From analysis of all significantly differentially expressed genes, we identified the following:
- SODC (superoxide dismutase) is strongly upregulated in coral hosting D compared to B; SODC is an NF-kB response gene; this is consistent with our expectations that coral hosting D would perform better under thermal/oxidative stress because increased SOD means an increased ability to deal with oxidative stress
- To a lesser extent, HSP110 may be downregulated in D compared to B, but this may be an effect of count value
- TNFSF10 and TRAF4.2 were identified as upregulated in one B (B.1) compared to D (i.e., downregulated in D), but again, this may be an effect of count size
